Add Arch Linux Usan 3.x/ build bot with `-fno-sanitize=function`

# Bug report

### Bug description:

Tests are broken for build bot Arch Linux Usan 3.x/4532 since the compilation fails with:

```
Objects/object.c:1071:16: runtime error: call to function long_hash through pointer to incorrect function type 'long (*)(struct _object *)'
/buildbot/buildarea/3.x.pablogsal-arch-x86_64.clang-ubsan/build/Objects/longobject.c:3618: note: long_hash defined here
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior Objects/object.c:1071:16 in
make: *** [Makefile:1691: Python/frozen_modules/getpath.h] Error 1
```

See for instance https://buildbot.python.org/#/builders/720/builds/1300.

### CPython versions tested on:

CPython main branch

### Operating systems tested on:

_No response_

--- a/master/custom/builders.py
+++ b/master/custom/builders.py
@@ -19,6 +19,7 @@ from custom.factories import (
     PGOUnixBuild,
     ClangUnixBuild,
     ClangUbsanLinuxBuild,
+    ClangUbsanFunctionLinuxBuild,
     ClangUnixInstalledBuild,
     SharedUnixBuild,
     SlowNonDebugUnixBuild,
@@ -310,6 +311,9 @@ UNSTABLE_BUILDERS_NO_TIER = [
 
     # riscv64 GCC
     ("riscv64 Ubuntu23", "onder-riscv64", SlowUnixInstalledBuild),
+
+    # Arch Usan Function
+    ("AMD64 Arch Linux Usan Function", "pablogsal-arch-x86_64", ClangUbsanFunctionLinuxBuild),
 ]
 
 

--- a/master/custom/factories.py
+++ b/master/custom/factories.py
@@ -322,6 +322,15 @@ class ClangUbsanLinuxBuild(UnixBuild):
     ]
     factory_tags = ["clang", "ubsan", "sanitizer"]
 
+class ClangUbsanFunctionLinuxBuild(UnixBuild):
+    buildersuffix = ".clang-ubsan-function"
+    configureFlags = [
+        "CC=clang",
+        "LD=clang",
+        "CFLAGS=-fsanitize=undefined -fno-sanitize=function -fsanitize-recover",
+        "--with-undefined-behavior-sanitizer",
+    ]
+    factory_tags = ["clang", "ubsan", "sanitizer"]
 
 class ClangUnixInstalledBuild(UnixInstalledBuild):
     buildersuffix = ".clang-installed"

